system/lib/wasmfs/syscalls.cpp

```diff
@@ -20,11 +20,50 @@
 #include <utility>
 #include <vector>
 #include <wasi/api.h>
+#include <unistd.h>
 
 extern "C" {
 
 using namespace wasmfs;
 
+// Copy the file specified by the pathname into JS.
+// Return a pointer to the JS buffer in HEAPU8.
+void* emscripten_wasmfs_read_file(char* path) {
+  struct stat file;
+  int err = 0;
+  err = stat(path, &file);
+  if (err < 0) {
+    emscripten_console_error(
+      "Fatal error in FS.readFile");
+    abort();
+  }
+  
+  size_t size = file.st_size;
+  uint8_t* result = (uint8_t*)malloc((size + sizeof(size)));
+  *(size_t*)result = size;
+  
+  int fd = open(path, O_RDONLY);
+  if (fd < 0) {
+    emscripten_console_error(
+      "Fatal error in FS.readFile");
+    abort();
+  }
+  err = pread(fd, result + sizeof(size), size, 0);
```

Comment:
It's probably worth asserting (for now) that the number of bytes read is the same as the original size with a TODO comment about making this more general and thread-proof in the future.